web/src/app/schedules/calendar-subscribe/CalendarSubscribeForm.js

```diff
@@ -0,0 +1,149 @@
+import React from 'react'
+import { PropTypes as p } from 'prop-types'
+import { FormContainer, FormField } from '../../forms'
+import { Grid, TextField } from '@material-ui/core'
+import { ScheduleSelect } from '../../selection'
+import MaterialSelect from '../../selection/MaterialSelect'
+import _ from 'lodash-es'
+
+export const reminderMinutesOptions = [
+  {
+    label: 'At time of shift',
+    value: '0',
+  },
+  {
+    label: '5 minutes before',
+    value: '5',
+  },
+  {
+    label: '10 minutes before',
+    value: '10',
+  },
+  {
+    label: '30 minutes before',
+    value: '30',
+  },
+  {
+    label: '1 hour before',
+    value: '60',
+  },
+  {
+    label: '1 day before',
+    value: '1440',
+  },
+]
+
+export default function CalendarSubscribeForm(props) {
+  return (
+    <FormContainer
+      disabled={props.loading}
+      errors={props.errors}
+      onChange={value => props.onChange(value)}
+      optionalLabels
+      value={props.value}
+    >
+      <Grid container spacing={2}>
+        <Grid item xs={12}>
+          <FormField
+            fullWidth
+            component={TextField}
+            name='name'
+            label='Subscription Name'
+            placeholder='My iOS Calendar'
+            required
+          />
+        </Grid>
+        <Grid item xs={12}>
+          <FormField
+            component={ScheduleSelect}
+            disabled={props.disableSchedField}
+            fullWidth
+            required
+            label='Schedule'
+            name='scheduleID'
+          />
+        </Grid>
+        {renderReminderMinutesFields()}
+      </Grid>
+    </FormContainer>
+  )
+
+  function renderReminderMinutesFields() {
+    let arr = _.get(props, 'value.reminderMinutes', []).slice()
+
+    const fields = []
+    // push 1 more field than there are values, max of 5 fields
+    // every field after the first is optional
+    for (let i = 0; i < arr.length + 1 && i < 5; i++) {
+      fields.push(
+        <Grid key={i} item xs={12}>
+          <FormField
+            fullWidth
+            component={MaterialSelect}
+            name={`reminderMinutes[${i}]`}
+            label={getAlarmLabel(i)}
+            required={i === 0}
+            options={reminderMinutesOptions}
+            mapValue={() => (arr[i] ? arr[i] : null)}
+            overrideOnChange={value => {
```

Comment:
We can do this, but this approach seemed straight forward enough and likely won't be used that often as this is a pretty edge case scenario. If you're uncomfortable having it as a prop though we can reconsider.

Reply:
I'd like to at least consider other options -- it will inevitably be used (see: copy/paste "because bugs") in the future to solve something else. The "just-override-everything" that is. It's not super portable since it depends on the current implementation of form components , (e.g. when we refactor the form stuff into hooks for example)

---

Let's discuss the details a bit more, just so I can get a better understanding of the problem and we can decide what to do from there.

Reply:
Valid point on that one lol, yeah that sounds good. The main reason was because it was still setting the value because it is wrapped in the parent’s onChange, and sometimes in the case of reminderMinutes we would want to have no change (or removing a val), but instead remove the field from the dom if cleared. But the parent onChange(onChange()) thing (psuedocode) was what was causing the issues. May be easier to show in person 😅

Reply:
Yeah, that will probably be easier. But just in case you take a look before we can chat: my current thought is maybe we can just update the `FormContainer` code to ALWAYS filter null values from arrays when an onChange handler returns null.

IIRC we have schedule rules, admin string list, and this as form fields that map to arrays, and in all 3 cases we'd never need a null value inside the array -- so maybe we can just "fix" it inside `FormContainer.js`.

Just pseudocode:
```js
// currently it boils down to a call like this, where null is the newValue
_.set(value, 'foo.bar[3]', null)

// If we are setting an array element and the new val is null
if (fieldName.endsWith(']') && newValue === null) {
// then just filter out that index
_.set(value, 'foo.bar', _.get('foo.bar', []).filter(idx=>idx!==3))
}
```